apps/compounds_tester/lib/compounds_tester_web/components/compounds/input.ex

```diff
@@ -1,57 +1,151 @@
 defmodule Compounds.Input do
   @moduledoc """
     The input component and it's sub-components.
-    TODO document the intention of this component.
-    We should make the current `input` component an `input_wrap` component, and create a seperate `input` component
-
-    TODO focus rings
+    ## TODO
+    * Clearable
+    * Password
+    * Input capture
   """
 
   use Phoenix.Component
+  alias Phoenix.LiveView.JS
 
   @doc """
     An input component.
 
     # Attributes:
-    * `class` - User defined TailwindCSS classes. This will be merge with the default class string.
-
     # Slots:
     * `label_right` - Label to be placed to the right of the input.
     * `label_left` - Label to be placed to the left of the input.
     * `icon_right` - Icon to be placed to the right of the input.
     * `icon_left` - Icon to be placed to the left of the input.
+    * `input_block_label` -
+    * `outer_wrapper` -
+    * `inner_wrapper` -
+    * `input_wrapper` -
+
+    # Attributes
+    * `class` -
+    * `type` -
+    * `placeholder` -
+    * `disabled?` -
+    * `read_only?` -
   """
   slot :label_right
```

Comment:
Woops, I targeted the wrong code block in the review.
Pretend that the review is attached to this section:
```elixir
slot :outer_wrapper do
    attr :id, :string, required: false
    attr :class, :string, required: false
  end

  slot :inner_wrapper do
    attr :id, :string, required: false
    attr :class, :string, required: false
  end

  slot :input_wrapper do
    attr :id, :string, required: false
    attr :class, :string, required: false
  end
```

Reply:
I think it's a matter of using them for the right thing. In the case of input it would be more verbose having to call these wrapper slots at functions instead. With named slots this also allows us to check if they're there and change things about the component conditionally which is important here for styling reasons. There are defiantly drawbacks to them but I don't think it's enough for a blanket "should not use them". You shouldn't _always_ use them either, of course.